File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/liveroom/core/UserManager.java
package com.trtc.uikit.livekit.liveroom.core;

import android.text.TextUtils;

import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMUserFullInfo;
import com.tencent.imsdk.v2.V2TIMValueCallback;
import com.trtc.uikit.livekit.common.utils.LiveKitLog;
import com.trtc.uikit.livekit.liveroom.core.listener.GetUserInfoCallback;
import com.trtc.uikit.livekit.liveroom.core.listener.GetUserListCallback;
import com.trtc.uikit.livekit.liveroom.data.UserInfo;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public class UserManager {
    private UserManager() {
    }

    private static class UserManagerHolder {
        private static final UserManager instance = new UserManager();
    }

    public static UserManager getInstance() {
        return UserManagerHolder.instance;
    }

    public void getUserInfo(String userId, GetUserInfoCallback listener) {
        LiveKitLog.info("UserManager getUserInfo:[userId:" + userId + "]");
        if (listener != null) {
            List<String> userList = new ArrayList<>();
            userList.add(userId);
            getUserList(userList, new GetUserListCallback() {
                @Override
                public void onSuccess(List<UserInfo> userList) {
                    if (userList != null && !userList.isEmpty()) {
                        listener.onSuccess(userList.get(0));
                    }
                }

                @Override
                public void onError(int errorCode, String errorMessage) {
                    listener.onError(errorCode, errorMessage);
                }
            });
        } else {
            LiveKitLog.error("UserManager getUserInfo:[Cancel]");
        }
    }

    public void getUserList(List<String> userList, GetUserListCallback listener) {
        LiveKitLog.info("UserManager getUserList:[userList:" + userList + "]");
        if (listener != null) {
            V2TIMManager.getInstance().getUsersInfo(userList, new V2TIMValueCallback<List<V2TIMUserFullInfo>>() {
                @Override
                public void onSuccess(List<V2TIMUserFullInfo> v2TIMUserFullInfoList) {
                    LiveKitLog.info("UserManager getUserList:[Success, v2TIMUserFullInfoList:"
                            + v2TIMUserFullInfoList + "]");
                    CopyOnWriteArrayList<UserInfo> userInfoList = new CopyOnWriteArrayList<>();
                    for (V2TIMUserFullInfo userFullInfo : v2TIMUserFullInfoList) {
                        if (!TextUtils.isEmpty(userFullInfo.getUserID())) {
                            UserInfo userInfo = new UserInfo();
                            userInfo.userId = userFullInfo.getUserID();
                            userInfo.name.set(userFullInfo.getNickName());
                            userInfo.avatarUrl.set(userFullInfo.getFaceUrl());
                            userInfoList.add(userInfo);
                        }
                    }
                    listener.onSuccess(userInfoList);
                }

                @Override
                public void onError(int error, String message) {
                    LiveKitLog.error("UserManager getUserList:[Error:" + error + ",message:" + message + "]");
                    listener.onError(error, message);
                }
            });
        } else {
            LiveKitLog.error("UserManager getUserList:[Cancel]");
        }
    }
}

File: iOS/TUILiveKit/Source/LiveRoom/Core/UserManager.swift
//
//  UserManager.swift
//  TUILiveKit
//
//  Created by WesleyLei on 2023/12/15.
//

import Foundation
import ImSDK_Plus
typealias GetUserInfoSuccessCallback = (_ userInfo: UserInfo?) -> Void
typealias GetUserListSuccessCallback = (_ userList: [UserInfo]) -> Void
public class UserManager {

    static func getUserInfo(userId: String, success: @escaping GetUserInfoSuccessCallback, fail: V2TIMFail? = nil) {
        getUserList(userList: [userId], success: { userList in
            success(userList.first)
        }, fail: fail)
    }

    static func getUserList(userList: [String], success: @escaping GetUserListSuccessCallback, fail: V2TIMFail? = nil) {
        LiveKitLog.info("\(#file)","\(#line)","getUserList:[userList:\(userList)]")
        V2TIMManager.sharedInstance().getUsersInfo(userList) { userFullInfoList in
            var list: [UserInfo] = []
            guard let userFullInfoList = userFullInfoList else {
                LiveKitLog.warn("\(#file)","\(#line)","getUserList:[userList:\(userList),userFullInfoList:0]")
                success(list)
                return
            }
            LiveKitLog.info("\(#file)","\(#line)","getUserList:[userList:\(userList),userFullInfoList:\(userFullInfoList.count)]")
            for userFullInfo in userFullInfoList {
                let user = UserInfo()
                user.userId = userFullInfo.userID
                user.avatarUrl.value = userFullInfo.faceURL
                user.name.value = userFullInfo.nickName
                list.append(user)
            }
            success(list)
        } fail: { code, message in
            LiveKitLog.error("\(#file)","\(#line)","getUserList:[userList:\(userList),code:\(code),message:\(String(describing: message))]")
            fail?(code, message)
        }
    }
}
